Example: Adding support for n differences

seabbs opened the pull request:
This is an example of a potential extension for adding n differences. The default 0 would be a model that had the mean of the cases (now fit in model) as mean of the GP. A first order would be a model that had a 0 growth rate as the mean of the GP. A second order would be a model that had r_t-1 as the mean at time t, this would be a logit version of the EpiNow2 default. Higher orders possible and maybe sensible but likely hard to fit (i.e would be assuming change in growth rate continues unless supported by data).

sbfnk commented on stan/inc2prev_antibodies.stan:
```diff
@@ -60,6 +61,7 @@ parameters {
   real<lower = 0> rho; // length scale of gp
   real<lower = 0> alpha; // scale of gp
   vector[M] eta; // eta of gp
+  real inc_init; // Initial infections
```
Should this be a vector of length `diff_order` where the first element is initial value of infections, second initial value of growth in infections etc.? See below

seabbs replied:
Yes agree if not comfortable with letting the GP having a slightly warped initial definition. I will do some testing and see how much the model complains as implemented to see if this is needed.

seabbs replied:
This does appear to really make much difference so I would probably suggest leaving as.

sbfnk commented on stan/inc2prev_antibodies.stan:
```diff
@@ -70,7 +72,7 @@ parameters {
 
 transformed parameters {
-  vector[t] gp; // value of gp at time t
+  vector[diff_order == 0 ? t : t-1] gp; // value of gp at time t
```
Should this be `t - diff_order`? Seems to me that if the GP prior applies to the differences in `r` then it would need an initial value of `r` otherwise I think it's assumed to be 0.

seabbs replied:
This is true, I think as written the first n terms of the GP will just have a slightly different definition to the rest (so when diff_order = 2, the first step will be the difference from 0 to the first r estimate and from then it will be r to r difference). The benefit of leaving as is is really just code simplicity and TBH I think it may not make much difference...

seabbs commented:
Some tweaks + switched to initialising the growth rates etc with their own parameters.

Some notes: 

1. All versions are slower than the version of the model on master. This is because all versions of the model are now scaled to a global mean which really they should always have been. When I first set the model up I looked at this and dropped due to run-time issues and it not making an apparent difference. 
2. Only the antibody model has been updated. I think we need to discuss what the aim of this is in order to decide what to support and what not to support. If wanting to support models with optional data requirements etc probably better do it in a single model vs supporting lots of sub-models with duplicate parts. 
3. As expected the growth rate model is quite a bit slower than the non-differenced model.

seabbs commented:
I've made the antibody model optional and renamed it to be the main model. I've also added date and gp approximation to the estimation script as if only interested in now we don't really need to estimate a whole year etc.

seabbs commented:
![growth](https://user-images.githubusercontent.com/20317649/149964795-f1b69c46-e72c-48d5-9eca-47ae3f15ee53.png)
![infections](https://user-images.githubusercontent.com/20317649/149964804-87b7113d-0cc2-4196-8999-d1d0bb68d0c9.png)
![pairs](https://user-images.githubusercontent.com/20317649/149964817-3075318c-1b26-45cf-b9f5-a0c081548094.png)
![prev-logit](https://user-images.githubusercontent.com/20317649/149964866-48592c0d-edc3-40e1-b0bd-039c5461fc91.png)
![prev](https://user-images.githubusercontent.com/20317649/149964872-263e27e3-2d6e-4c73-b63b-853e448422f4.png)
![Rt](https://user-images.githubusercontent.com/20317649/149964880-ad19cd37-5b37-480c-a035-4014a5d8681c.png)

seabbs commented:
@sbfnk it would be good to know how your branch has diverged from this and/or what updates need to be made to support it? I've built subsequent updates on top of it so would be good to make sure makes sense.
